Add Machine ID Kubernetes and Apps guides (#14976)
gravitational/teleport: docs/pages/machine-id/guides/applications.mdx

---
title: Machine ID with Application Access
description: Accessing web applications with Machine ID
---

In this guide, we will demonstrate how to use Machine ID to access a web
application protected by Teleport from an automated service.

With Machine ID, Teleport issues short-lived certificates, tied to a machine
identity, that can be rotated, audited, and managed with the same access
controls that Teleport provides for human users.

## Prerequisites

<ScopedBlock scope={["cloud"]}>
You will need Teleport Cloud 10.1.0 or later.
</ScopedBlock>
<ScopedBlock scope={["oss","enterprise"]}>
You will need to be running the Teleport version 10.1.0 or later.
</ScopedBlock>

If you have not already connected your application to Teleport, follow
the [Application Access Getting Started Guide](../../application-access/getting-started.mdx).

If you're not already familiar with Machine ID, follow the
[Getting Started Guide](../getting-started.mdx) to familiarize yourself with
Machine ID. You'll also need `tctl` access to initially configure the bot.

(!docs/pages/includes/tctl.mdx!)

> Lastly, ensure `tbot` is installed on your Machine ID client system. Refer to

Author: I guess I'm not sure what that would look like beyond what the installation page already provides? We're really just referring to the `tbot` binary itself which is included in the `.deb` / `.rpm` / `.tar.gz` / etc packages. We do walk through the rest of the install steps in the guide (like the systemd service install).

I've tried to clarify the scope of what readers need from the install doc (just the binary), so let me know if its sufficient?

Service, it's also possible to open a local proxy to the app using
`tbot proxy app ...` in situations where the public proxy's app endpoints are
unavailable:

<ScopedBlock scope={["cloud"]}>
```code
$ tbot proxy --destination-dir=/opt/machine-id --proxy=example.teleport.sh:443 app --port=1234 grafana-example
```
</ScopedBlock>
<ScopedBlock scope={["oss","enterprise"]}>
```code
$ tbot proxy --destination-dir=/opt/machine-id --proxy=teleport.example.com:443 app --port=1234 grafana-example
```
</ScopedBlock>

You may now connect to the app via a local proxy at `http://localhost:1234`:

```code
$ curl --user admin:admin http://localhost:1234/api/users
```

Note that any clients with access to `localhost` may connect to this proxy
without additional authentication, so this should only be run in trusted
environments.
</Admonition>

## Troubleshooting

### Client application requires certificates with standard extensions

If your client application requires TLS certificates with a specific file
extension, you may also enable the `tls` config template in `tbot.yaml`:

```yaml
destinations:
  - directory: /opt/machine-id
    app: grafana-example

    configs:

>       - tls

Reviewer: What are the possible values of the `configs` field? Are these arbitrary strings that `tbot` uses to generated filenames?

Author: I've added a note explaining what this is and linked out to the config reference which documents other value entries.

```

This will generate `tls.crt` and `tls.key` inside `/opt/machine-id`
with identical content to the certificate files listed above.

Be sure to re-run `tbot init ...` to configure permissions for the new file if
config templates are changed.

### Clients are redirected to the Teleport login page

As with human users, scripted clients will be redirected to the Teleport login

> page when attempting to access an app through the public proxy without valid

Reviewer: ```suggestion
page when attempting to access an app through the Teleport Proxy Service without valid
```

---
title: Machine ID with Application Access
description: Accessing web applications with Machine ID
---

In this guide, we will demonstrate how to configure an automated service to use Machine ID to access a web application protected by Teleport.

With Machine ID, Teleport issues short-lived certificates, tied to a machine
identity, that can be rotated, audited, and managed with the same access
controls that Teleport provides for human users.

## Prerequisites

<ScopedBlock scope={["cloud"]}>
You will need Teleport Cloud 10.1.0 or later.
</ScopedBlock>
<ScopedBlock scope={["oss","enterprise"]}>

> You will need to be running Teleport version 10.1.0 or later.

Reviewer: ```suggestion
You will need a running Teleport cluster version 10.1.0 or later.
```
To distinguish running a cluster from, say, having a single `teleport` binary running on a host.

</ScopedBlock>

If you have not already connected your application to Teleport, follow
the [Application Access Getting Started Guide](../../application-access/getting-started.mdx).

If you're not already familiar with Machine ID, follow the
[Getting Started Guide](../getting-started.mdx) to familiarize yourself with
Machine ID. You'll also need `tctl` access to initially configure the bot.

(!docs/pages/includes/tctl.mdx!)

> Lastly, ensure the `tbot` binary is installed on your Machine ID client system.

Reviewer: I think we should be more specific about what the client system entails. Is this a VM with access to the Auth Service or Proxy Service and the application you want Machine ID to access? Would a user be able to follow this guide with the `tbot` binary installed on their local machine?

Author: I've reworked this to hopefully better explain what a client system is. I didn't say it explicitly, but it is perfectly fine to run it on your local machine to try things out. I wasn't sure if that would create more confusion, though.

```yaml
auth_server: "example.teleport.sh:443"
onboarding:
  join_method: "token"
  token: "00000000000000000000000000000000"
  ca_pins:
  - "sha256:1111111111111111111111111111111111111111111111111111111111111111"
storage:
  directory: /var/lib/teleport/bot
destinations:
  - directory: /opt/machine-id
    app: grafana-example
```
</ScopedBlock>
<ScopedBlock scope={["oss","enterprise"]}>
```yaml
auth_server: "teleport.example.com:443"
onboarding:
  join_method: "token"
  token: "(=presets.tokens.first=)"
  ca_pins:
  - "(=presets.ca_pin=)"
storage:
  directory: /var/lib/teleport/bot
destinations:
  - directory: /opt/machine-id
    app: grafana-example
```
</ScopedBlock>

Be sure to configure the `token` and `ca_pins` fields to match the output from
`tctl bots add ...`, and set `app` to match the desired name as shown in
`tsh apps ls`. For this example, we're using the `grafana-example` app.

Machine ID also allows you to use Linux ACLs to control access to certificates
on disk. You will use this to ensure only your application has access to the
short-lived certificates Machine ID uses.

We'll work with the assumption you will be running Machine ID as the Linux user

> `teleport` and your application as the Linux user `app`. Create and initialize

Reviewer: > We'll work with the assumption you will be running Machine ID as the Linux user
> `teleport` and your application as the Linux user `app`.

I think we should make this more explicit that "your application" refers to the app you're using to access an upstream application protected by Teleport, not the upstream application itself (assuming that's correct).

Later, the guide says this application is an "automated service." I think that term works, but we should also clarify earlier in the guide that Machine ID will be providing credentials to an automated service that is connecting to your application.

```

Note that any clients with access to `localhost` may connect to this proxy
without additional authentication, so this should only be run in trusted
environments.
</Admonition>

## Troubleshooting

### Client application requires certificates with standard extensions

If your client application requires TLS certificates with a specific file
extension, you may also enable the `tls` config template in `tbot.yaml`:

```yaml
destinations:
  - directory: /opt/machine-id
    app: grafana-example

    configs:
      - tls
```

This will generate `tls.crt` and `tls.key` inside `/opt/machine-id`
with identical content to the certificate files listed above.

Be sure to re-run `tbot init ...` to configure permissions for the new file if
config templates are changed.

### Clients are redirected to the Teleport login page

As with human users, scripted clients will be redirected to the Teleport login
page when attempting to access an app through the Teleport Proxy Service without valid
credentials.

Ensure the bot's certificates have not expired and that the client application
has been configured to use both the client certificate and key.

## Next steps

> - Review the 

Reviewer: We'll need to complete this

Author: Whoops, clearly my brain wasn't working. Will fix.

configured above. Be sure to note the bot join token and CA PIN.

Next, on the bot node, create a Machine ID configuration file at
`/etc/tbot.yaml`:

<ScopedBlock scope={["cloud"]}>
```yaml
auth_server: "example.teleport.sh:443"
onboarding:
  join_method: "token"
  token: "00000000000000000000000000000000"
  ca_pins:
  - "sha256:1111111111111111111111111111111111111111111111111111111111111111"
storage:
  directory: /var/lib/teleport/bot
destinations:
  - directory: /opt/machine-id
    app: grafana-example
```
</ScopedBlock>
<ScopedBlock scope={["oss","enterprise"]}>
```yaml
auth_server: "teleport.example.com:443"
onboarding:
  join_method: "token"
  token: "(=presets.tokens.first=)"
  ca_pins:
  - "(=presets.ca_pin=)"
storage:
  directory: /var/lib/teleport/bot
destinations:
  - directory: /opt/machine-id
    app: grafana-example
```
</ScopedBlock>

Be sure to configure the `token` and `ca_pins` fields to match the output from
`tctl bots add ...`, and set `app` to match the desired name as shown in
`tsh apps ls`. For this example, we're using the `grafana-example` app.

Machine ID also allows you to use Linux ACLs to control access to certificates

> on disk. You will use this to ensure only your application has access to the

Reviewer: ```suggestion
on disk. You will use this to ensure only your automated service has access to the
```
To distinguish this from the upstream application

will be redirected to the Teleport login page when attempting to access the app.

<Admonition type="note" title="Proxy Note">
While it's usually best to access apps through the public Teleport Proxy
Service, it's also possible to open a local proxy to the app using
`tbot proxy app ...` in situations where the Teleport Proxy Service's app endpoints are
unavailable:

<ScopedBlock scope={["cloud"]}>
```code
$ tbot proxy --destination-dir=/opt/machine-id --proxy=example.teleport.sh:443 app --port=1234 grafana-example
```
</ScopedBlock>
<ScopedBlock scope={["oss","enterprise"]}>
```code
$ tbot proxy --destination-dir=/opt/machine-id --proxy=teleport.example.com:443 app --port=1234 grafana-example
```
</ScopedBlock>

You may now connect to the app via a local proxy at `http://localhost:1234`:

```code
$ curl --user admin:admin http://localhost:1234/api/users
```

Note that any clients with access to `localhost` may connect to this proxy
without additional authentication, so this should only be run in trusted
environments.
</Admonition>

## Troubleshooting

### Client application requires certificates with standard extensions

> If your client application requires TLS certificates with a specific file

Reviewer: ```suggestion
If your automated service requires TLS certificates with a specific file
```
To ensure we keep terms consistent
